### DISCUSSION TITLE
calling stack

### DISCUSSION THREAD

**Discussion by nobanzai** [gerbera/gerbera#3270] [General]
Hi again,

I have still problems to understand, which functions are called in which order, probably also because of some inconsistencies in the the docs.
In "Best practices for testing" we find:
`Do not modify the provided common.js file. Override its functions in a separate file and let Gerbera load it by specifying it in custom-script in config.xml.`
But setting custom-script is deprecated as we can read in [https://docs.gerbera.io/en/stable/config-import.html#custom-script](url).
Are the following assumptions correct?
1.  The starting point for importing a file is **always** /usr/share/gerbera/js/import.js? But why is it called 23 times in my environment on every restart of gerbera? There were no new files in the directory structure at this point.
2. Is it possible to replace import.js by my own function and if so how, after custom-script is deprecated? Would it even make sense?
3. In import.js itself the first step is to call importItem, which in turn calls the various import<mediatype> functions. But from where is a script called, that is configured in the import-function section of config.xml, for examle importAudioKarl in the docs?

It would be great, if there would be some flowchart, that could visualize the different ways and the order the functions are called.

Thx!

Bye.
Michael.

**Comment by KarlStraussberger** [comment]
> `Do not modify the provided common.js file. Override its functions in a separate file and let Gerbera load it by specifying it in custom-script in config.xml.`
> But setting custom-script is deprecated as we can read in [https://docs.gerbera.io/en/stable/config-import.html#custom-script](url).

It's replaced by `script-folder/custom` 

> Are the following assumptions correct?
> 1.  The starting point for importing a file is **always** /usr/share/gerbera/js/import.js? But why is it called 23 times in my environment on every restart of gerbera? There were no new files in the directory structure at this point.

The starting point is the function specified as `import-function` regardless of the file. The script file is loaded and parsed on startup to get the function cached.

> 2. Is it possible to replace import.js by my own function and if so how, after custom-script is deprecated? Would it even make sense?

You only have to provide the new function. `import.js` is just containing the wrapper code for compatibility.

> 3. In import.js itself the first step is to call importItem, which in turn calls the various import<mediatype> functions. But from where is a script called, that is configured in the import-function section of config.xml, for examle importAudioKarl in the docs?

The switch can be regarded as legacy.

**Reply by nobanzai** [reply]
Obviously I am not really able to use the right words to express my problems to understand the structure.

> > `Do not modify the provided common.js file. Override its functions in a separate file and let Gerbera load it by specifying it in custom-script in config.xml.`
> > But setting custom-script is deprecated as we can read in [https://docs.gerbera.io/en/stable/config-import.html#custom-script](url).
> 
> It's replaced by `script-folder/custom`

Yes, thas is, what I use:
```
    <script-folder>
        <common>/usr/share/gerbera/js</common>
        <custom>/home/gerbera/jscript</custom>
      </script-folder>

```
> 
> > Are the following assumptions correct?
> > 
> > 1. The starting point for importing a file is **always** /usr/share/gerbera/js/import.js? But why is it called 23 times in my environment on every restart of gerbera? There were no new files in the directory structure at this point.
> 
> The starting point is the function specified as `import-function` regardless of the file. The script file is loaded and parsed on startup to get the function cached.

But why is the file "run" for 23 times - I added a print2 statement and it prints out its text 23 times?
Is this also for compatibility?

> 
> > 2. Is it possible to replace import.js by my own function and if so how, after custom-script is deprecated? Would it even make sense?
> 
> You only have to provide the new function. `import.js` is just containing the wrapper code for compatibility.

Ok, so
```
      <import-function>
        <audio-file>importAudio_mh001</audio-file>
        ...
      </import-function>
```
is sufficient.

> 
> > 3. In import.js itself the first step is to call importItem, which in turn calls the various import functions. But from where is a script called, that is configured in the import-function section of config.xml, for examle importAudioKarl in the docs?
> 
> The switch can be regarded as legacy.

Don't understand this one - who calls my importAudio_mh001?

Sorry for being that pushing, but I really would like to understand it completely.

Thx and bye.
Michael

**Reply by KarlStraussberger** [reply]
1. There are several internal "contexts". In order to avoid side effects by global variables in a context, each context is initiated separately, hence you 23 calls.
2. Yes
3. The function is directly call from the C++ code (it's an entrypoint). The call provides all parameters and reads the response values. Until some time ago we always executed the whole script which was very restrictive and did not allow to modularise the code.

**Comment by nobanzai** [comment]
Ok, got it.
Thx a lot!

**Reply by KarlStraussberger** [reply]
You're welcome
